### ISSUE TITLE
optimize/convex/lp: Simplex should panic if m > n

### ISSUE THREAD

**Issue by jjhbw** [gonum/gonum#450]
Thank you for this amazing package. Really nice work. I has become my go-to library for when I need to 'rescue' scientific code from R or matlab.
I am currently implementing a simple integer linear programming routine on top of simplex, and came across the following error. 

In short: `lp.Simplex` tells me my input matrix `A` is singular but I don't think it actually is.

### What are you trying to do?
Solve the linear problem specified below.

### What did you do?
```go
func TestSingularityReproduction(t *testing.T) {
	A := mat.NewDense(16, 14, []float64{
		1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 0, 0,
		0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
		0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
		1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
		0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
		0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
		0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
		0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0,
		0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
		0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
	})

	b := []float64{1, 2, 1, 1, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1}

	c := []float64{-0, -0, -0, -0, -0, -0, -1, 0, 0, 0, 0, 0, 0, 0}

	_, _, err := lp.Simplex(c, A, b, 0, nil)
	if err != nil {
		t.Error(err)
		// lp: A is singular
	}
}
```

### What did you expect to happen?
I would expect a feasible solution.

### What actually happened?
I received the `lp: A is singular` error.
Which is produced by:  https://github.com/gonum/gonum/blob/6b03bc22e15aece534f0b4fb96cf6c8f1d885cd0/optimize/convex/lp/simplex.go#L496

### What version of Go and Gonum are you using?
Go:
```
go version go1.10 darwin/amd64
```

Gonum:
052ff404ed7bcbcf5c326f45dd2e8021cef6af9f

### Does this issue reproduce with the current master?
Yes. ( with 6b03bc22e15aece534f0b4fb96cf6c8f1d885cd0)

**Comment by btracey** [comment]
Your matrix is singular. For instance, row_0 - row_1 + row_6 = 0 . In particular, this error is saying that there does not exist a set of 14 columns that are independent. If you can find such a subset, please reopen.

**Comment by btracey** [comment]
Or, if you have a good argument for why you don't think your matrix is singular

**Comment by jjhbw** [comment]
Oops, thanks! Looks like the function I wrote to find cases like these is lacking.

**Comment by btracey** [comment]
General case: Why is it strange to say a non-square matrix is singular? It's saying that the rank is less than min(r, c). It means you can't run Solve, etc.

Specific case: I think there is a bug, because this matrix does have full rank.

@jjhbw My apologies, looking in to it.

**Comment by vladimir-ch** [comment]
It depends on how you define "singular". For me singular matrix is equivalent to noninvertible matrix is equivalent to zero determinant is equivalent to rank less than n (all notions for square matrices). For you it's rank less than min(r,c).

**Comment by btracey** [comment]
Okay, reviewing the code, this code, and the Simplex  algorithm more generally, assumes that `m < n`. The simplex works by choosing `m` active constraints, iteratively solving `A x = b`, and adjusting those constraints. There are clearly feasible linear programs that have more constraints than variables, but I think that's outside the scope of `Simplex` (just as LPs in non-standard form are outside the scope of `Simplex` but we have a `Convert` function). 

The immediate answer is to panic with a clear message. 

The more general answer is to provide a pre-processing step that can deal with these redundancies. One has to both find the redundant equations, but also which of those is the least restrictive. I agree this is possible in principal, but I don't know the algorithm, and the basic tutorials of Simplex I have read do not discuss this.

**Comment by btracey** [comment]
Also @jjhbw , note that while `lp` does work (though without the bells and whistles of robust solvers, as this issue makes clear), at the moment it's significantly slower than other solvers especially for large, sparse programs ( @vladimir-ch  and I are working on it). If you have access to Gurobi, I have a simple package for building LPs and then interfacing with various solvers. github.com/btracey/lpwrap . It needs examples, but I've found it relatively straightforward to use.

**Comment by btracey** [comment]
Fixed by #451 (kind of ... feel free to open an issue about a pre-processing routine)

**Comment by jjhbw** [comment]
Awesome! I agree that it is more intuitive this way.
Also thanks for pointing me to the gurobi wrapper. Will definitely look into that when i need to tap into a more feature-rich solver suite from Go. For now i like the Gonum simplex implementation precisely for the portability that using pure Go provides.
